dvc/ignore.py

```diff
@@ -79,13 +88,53 @@ def ignore(self, path, is_dir):
         return result
 
     def __hash__(self):
-        return hash(self.ignore_file_path)
+        return hash(self.dirname + ":" + "\n".join(self.pattern_list))
 
     def __eq__(self, other):
         if not isinstance(other, DvcIgnorePatterns):
             return NotImplemented
+        return (self.dirname == other.dirname) & (
+            self.pattern_list == other.pattern_list
+        )
+
+    def __bool__(self):
+        return bool(self.pattern_list)
+
 
-        return self.ignore_file_path == other.ignore_file_path
+class DvcIgnorePatternsTrie(DvcIgnore):
+    trie = None
+
+    def __init__(self):
+        if self.trie is None:
+            self.trie = StringTrie(separator=os.sep)
+
+    def __new__(cls, *args, **kwargs):
+        if not hasattr(DvcIgnorePatterns, "_instance"):
+            if not hasattr(DvcIgnorePatterns, "_instance"):
+                DvcIgnorePatterns._instance = object.__new__(cls)
+        return DvcIgnorePatterns._instance
+
+    def __call__(self, root, dirs, files):
+        ignore_pattern = self[root]
+        if ignore_pattern:
+            return ignore_pattern(root, dirs, files)
+        return dirs, files
+
+    def __setitem__(self, root, ignore_pattern):
+        base_pattern = self[root]
+        common_dirname, merged_pattern = merge_patterns(
+            base_pattern.dirname,
+            base_pattern.pattern_list,
+            ignore_pattern.dirname,
+            ignore_pattern.pattern_list,
+        )
```

Comment:
So we collect everything as we go in. We still leave partial results along the way. This means the overall structure will take quadratic memory depending on tree depth. Is this ok?

Reply:
@Suor 
Yes, the inherited pattern would be duplicated several times. If not so, we would merge all of them at the file's checking. Space or speed? It depends on which is more important, I think.